app/src/main/java/io/github/jbellis/brokk/SessionManager.java

```diff
@@ -161,6 +163,65 @@ public void moveSessionToUnreadable(UUID sessionId) {
                 logger.error("Error moving history zip for session {} to unreadable: {}", sessionId, e.getMessage());
             }
         });
+        try {
+            future.get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /** Detailed report of sessions quarantined during a scan. */
+    public record QuarantineReport(
+            Set<UUID> quarantinedSessionIds, List<String> quarantinedFilesWithoutUuid, int movedCount) {}
+
+    /**
+     * Scans the sessions directory for all .zip files, treating any with invalid UUID filenames or missing/invalid
+     * manifest.json as unreadable and moving them to the 'unreadable' subfolder. Also attempts to load each valid-UUID
+     * session's history to exercise migrations. Returns a report detailing what was quarantined.
+     *
+     * <p>This runs synchronously; intended to be invoked from a background task.
+     */
+    public QuarantineReport quarantineUnreadableSessions(IContextManager contextManager) {
+        int moved = 0;
+        var quarantinedIds = new HashSet<UUID>();
+        var quarantinedNoUuid = new ArrayList<String>();
+
+        try (var stream = Files.list(sessionsDir)) {
+            for (Path zipPath :
+                    stream.filter(p -> p.toString().endsWith(".zip")).toList()) {
+                var maybeUuid = parseUuidFromFilename(zipPath);
+                if (maybeUuid.isEmpty()) {
+                    // Non-UUID filenames are unreadable by definition.
+                    moveZipToUnreadable(zipPath, null);
+                    quarantinedNoUuid.add(zipPath.getFileName().toString());
+                    moved++;
+                    continue;
+                }
+
+                var sessionId = maybeUuid.get();
+                var info = readSessionInfoFromZip(zipPath);
+                if (info.isEmpty()) {
+                    quarantinedIds.add(sessionId);
+                    moved++;
+                    continue;
+                }
```

Comment:
These aren't moved to `unreadable` folder here, only the counter is incremented.